### ISSUE TITLE
TypeError: Cannot read property 'split' of undefined

### ISSUE THREAD

**Issue by ig-5g** [garden-io/garden#2003]
### Current Behavior

`garden delete env` or `garden delete service <service-name>` result in:

```
[2020-08-12T17:49:38.810Z] TypeError: Cannot read property 'split' of undefined
    at makeErrorMsg (/snapshot/project/garden-service/tmp/dist/build/src/util/util.js:0)
    at Object.<anonymous> (/snapshot/project/garden-service/tmp/dist/build/src/util/util.js:0)
    at Generator.throw (<anonymous>)
    at rejected (/snapshot/project/garden-service/tmp/dist/build/src/util/util.js:0)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

This is a regression in garden `0.12.x`. We don't see this error in `0.11.x` 

The error is not platform-specific. We get the same error on MacOS, CentOS 7 and Windows 10.

### Expected behavior

`garden delete env` and `garden delete service` work cleanly in `0.12.x`

### Reproducible example

We are still working on creating a minimal reproducible `garden.yaml`.  We need help with the error message -- see below.

### Workaround

We could not find one. We pinned our garden version to `0.11.14`

### Suggested solution(s)

There are two issues here. The `split` error comes from this line in `util.ts`

```
export function makeErrorMsg({
   ...
   const lines = output.split("\n")
   ...
}
```

The code dereferences undefined `output`. This bug masks the original error.

It'd be nice to fix the undefined `output` issue in the upcoming `0.12.x` release. Once it's fixed, we hope to see more clues about the root cause of the problem.

### Your environment

```
$ garden version
0.12.2
```

Note: we started getting the error in `0.12.0`. The regression happened between `0.11.14` and `0.12.0`

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6", GitCommit:"dff82dc0de47299ab66c83c626e08b245ab19037", GitTreeState:"clean", BuildDate:"2020-07-15T16:58:53Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.9+IKS", GitCommit:"dbe4e8a5c169e0a6a8a9b7b183561fa61e161985", GitTreeState:"clean", BuildDate:"2020-07-16T01:30:57Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

`docker version`
N/A - we run docker builds in a different environment. docker is not present in the environment where we run garden.

**Comment by edvald** [comment]
Thanks @ig-5g! As you suggested, I'll start by submitting a PR for the undefined error, then we can hopefully get at the root cause. Watch this space :)

**Comment by ig-5g** [comment]
@edvald -- thanks for the fix in `v0.12.3`. I re-ran the same use case and got more details about the root cause.

`garden delete service my-service` results in:

```
Failed resolving one or more modules:

module-1: Command "git diff-index --name-only HEAD /home/ig-5g/my-project/helm" failed with code undefined:

Here's the full output:


module-2: Command "git ls-files -s --others --exclude /home/ig-5g/my-project/helm/.garden --exclude-per-directory .gardenignore /home/ig-5g/my-project/helm" failed with code undefined:

Here's the full output:

```

`git diff-index` error repeats a few more times for other modules.

**Every module in question is a local Helm chart.** We do not specify `repositoryUrl` or `repo` properties in the module definitions.

Project structure:

```
my-project
├── .git
|
├── helm
    ├── garden.yml
    ├── helm chart 1
    ├── helm chart 2
    ├── ...
```

As I mentioned before, the same structure works fine with garden `v0.11.14`

A couple of snippets from `error.log`

```
[2020-09-01T00:27:32.926Z] Error: Command "git diff-index --name-only HEAD /home/ig-5g/my-project/helm" failed with code undefined:

Here's the full output:


    at Process.ChildProcess._handle.onexit (internal/child_process.js:267:19)
    at onErrorNT (internal/child_process.js:469:16)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
Error Details:
errno: EMFILE
code: EMFILE
syscall: spawn git
path: git
spawnargs: []
originalMessage: spawn git EMFILE
shortMessage: >-
  Command failed with EMFILE: git diff-index --name-only HEAD
  /home/ig-5g/my-project/helm

  spawn git EMFILE
command: git diff-index --name-only HEAD /home/ig-5g/my-project/helm
failed: true
timedOut: false
isCanceled: false
killed: false


[2020-09-01T00:27:35.944Z] Error: Command "git ls-files -s --others --exclude /home/ig-5g/my-project/helm/.garden --exclude-per-directory .gardenignore
 /home/ig-5g/my-project/helm" failed with code undefined:

Here's the full output:


    at Process.ChildProcess._handle.onexit (internal/child_process.js:267:19)
    at onErrorNT (internal/child_process.js:469:16)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
Error Details:
errno: EMFILE
code: EMFILE
syscall: spawn git
path: git
spawnargs: []
originalMessage: spawn git EMFILE
shortMessage: >-
  Command failed with EMFILE: git ls-files -s --others --exclude
  /home/ig-5g/my-project/helm/.garden --exclude-per-directory
  .gardenignore /home/ig-5g/my-project/helm

  spawn git EMFILE
command: >-
  git ls-files -s --others --exclude
  /home/ig-5g/my-project/helm/.garden --exclude-per-directory
  .gardenignore /home/ig-5g/my-project/helm
failed: true
timedOut: false
isCanceled: false
killed: false
```

**Comment by edvald** [comment]
Aha, I see. `EMFILE` means there are too many open files on your system, which can happen on *nix systems if your limits are too low. Which OS are you running?

**Comment by ig-5g** [comment]
@edvald  -- I am testing on CentOS 7.7

I tried increasing `open files` limit from 1024 to 4096 to 10000 to 20000. Same EMFILE error.

I then took the bazooka out and set the limit to the kernel limit:

```
$ cat /proc/sys/fs/file-max
183634
```

```
$ cat /etc/security/limits.conf | grep nofile
* hard nofile 183634
* soft nofile 183634
```

```
$ ulimit -Hn
183634

$ ulimit -Sn
183634
```

With this, I got a different failure.

```
$ garden delete service my-service
ℹ providers                 → Getting status...
✔ providers                 → Getting status... → Cached
   ℹ Run with --force-refresh to force a refresh of provider statuses.

<--- Last few GCs --->
in[25620:0x40a4e40]    27478 ms: Mark-sweep 748.7 (787.6) -> 747.6 (788.1) MB, 46.0 / 1.1 ms  (+ 877.5 ms in 149 steps since start of marking, biggest step 19.2 ms, walltime since start of marking 940 ms) (average mu = 0.289, current mu = 0.261) finalize in[25620:0x40a4e40]    28309 ms: Mark-sweep 748.2 (788.1) -> 747.2 (794.6) MB, 29.6 / 1.0 ms  (+ 721.3 ms in 69 steps since start of marking, biggest step 33.5 ms, walltime since start of marking 761 ms) (average mu = 0.231, current mu = 0.163) finalize inc

<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x12cb9d9]
Security context: 0x359e160408d1 <JSObject>
    1: read [0x2949003be769] [pkg/prelude/bootstrap.js:~560] [pc=0x22834cd24cff](this=0x017524642689 <Object map = 0x159439c388d9>,29058,0x3ce4eef886d1 <Uint8Array map = 0x34a10d12aa59>,0,65536,0x1d3dd7e404b1 <undefined>)
    2: arguments adaptor frame: 6->5
    3: _read [0x3b8a85eed3e1] [internal/fs/streams.js:~144] [pc=0x22834ce4e9a4](this=0x175df74a8351 <ReadS...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x9cc540 node::Abort() [garden]
 2: 0x9ce6c0 node::OnFatalError(char const*, char const*) [garden]
 3: 0xaeb86e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [garden]
 4: 0xaebbe7 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [garden]
 5: 0xc809c5  [garden]
 6: 0xc8135f v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [garden]
 7: 0xc8c32b v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [garden]
 8: 0xc8d1b7 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [garden]
 9: 0xc8e6df v8::internal::Heap::HandleGCRequest() [garden]
10: 0xc3df77 v8::internal::StackGuard::HandleInterrupts() [garden]
11: 0xf6feb6 v8::internal::Runtime_StackGuard(int, unsigned long*, v8::internal::Isolate*) [garden]
12: 0x12cb9d9  [garden]
Aborted
```

**Comment by edvald** [comment]
I see. So, first off this error shouldn't manifest this way, so I'll for sure qualify this as a bug on our side. I'm working on some improvements when handling these situations, but it appears you have a huge number of files in your project, and that some of them need to be explicitly ignored.

This section here covers how to go about that: https://docs.garden.io/using-garden/configuration-overview#including-excluding-files-and-directories

I suspect that you have some large dependency folders in the project. Normally that just results in slow performance, but in your case it seems to blow up the process, so I'll dig into that. But do take a look at that docs section and see if there are folders you can ignore from your project, and let me know if that does the trick!

**Comment by ig-5g** [comment]
@edvald -- Some project metrics just as an FYI. The number of files doesn't strike me as huge.

Number of folders in the project root:

```
$ find . -mindepth 1 -maxdepth 1 -type d | wc -l
29
```

Of 29 folders:
* 27 folders are modules. Each module is a simple Helm chart with 6-7 files per chart.
* 2 other folders are tiny: 6 files in total


Total number of folders under the project root:

```
$ find . -mindepth 1 -type d | wc -l
56
```

Total number of files under the project root:

```
$ find . -mindepth 1 -type f | wc -l
160
```

`.garden` folder metrics after full deployment:

```
cd .garden

# count directories
$ find . -type d | wc -l
1121

# count files
$ find . -type f | wc -l
2832
```

There is really not much we can put on ignore.

**Comment by edvald** [comment]
That's not huge at all, you are right, you should have no issues at all with those numbers. I wonder what else could cause this then.
Perhaps some recursion or something circular that we're not detecting. Do you have any directory symlinks in there? How is your project/module structure? We have circular dependency detection in several places but I'm wondering if that's not working in some specific case.

**Comment by edvald** [comment]
Thank you btw for all the detailed info, this is all super helpful!

**Comment by ig-5g** [comment]
I fixed the issue. I can run `garden delete service` or `garden delete env` without errors, with `open file` limit back to 1024.

I will write up the details tomorrow. It's getting a bit late here in NA.

**Comment by ig-5g** [comment]
## TL;DR

To fix the issue, I reduced the scope of `include` statements in the module definitions.

## Details

To recap, out project structure looks like this:

```
my-project
├── .git
|
└── helm
    ├── garden.yml
    ├── helm chart 1
    ├── helm chart 2
    ├── ...
```

We have one `garden.yml` for the entire project. It includes the project definition and all the modules.

Before the fix, all module specs looked like this:

```
kind: Module
type: helm
name: some-service
chartPath: some-service/
include: ["**"]
...
```

As you can see, **each helm module included the entire project tree.**

To fix the issue, I reduced the scope of include statements to include only what's necessary:

```
kind: Module
type: helm
name: some-service
chartPath: some-service/
include:
  - some-service/**/*
  - value-file-foo.yaml
  - value-file-bar.yaml
...
```

`.garden` folder metrics

before:
```
Number of directories: 1121
Number of files: 2832
```

after:
```
Number of directories: 166
Number of files: 162
```

@edvald -- you've been very helpful. Thank you very much.

**Comment by thsig** [comment]
Has this one been closed by [`
improvement: explicitly catch EMFILE errors with better error message`](https://github.com/garden-io/garden/commit/37975c86bca6d3aad81e1d422ab4ac0d60a20a73)?

**Comment by ig-5g** [comment]
@thsig -- yes, feel free to close the issue.
